db/migrate/20231018192110_add_index_to_webauthn_credentials_user_id_nickname.rb

```diff
@@ -0,0 +1,49 @@
+# frozen_string_literal: true
+
+class AddIndexToWebauthnCredentialsUserIdNickname < ActiveRecord::Migration[7.0]
+  disable_ddl_transaction!
+
+  def up
+    add_index_to_table
+  rescue ActiveRecord::RecordNotUnique
+    remove_duplicates_and_reindex
+  end
+
+  def down
+    remove_index_from_table
+  end
+
+  private
+
+  def remove_duplicates_and_reindex
+    deduplicate_records
+    reindex_records
+  rescue ActiveRecord::RecordNotUnique
+    retry
+  end
+
+  def reindex_records
+    remove_index_from_table
```

Comment:
If `add_index_to_table` fails with an exception, then the index is not created. `remove_duplicates_and_reindex` is called, which will call this, which will remove an index that does not exists.

I am missing something here?

Reply:
Without that line, we get a failure about trying to create an already existing index, so we need to remove it first.

This was a while ago so I'd need a refresher on the details and I also just copied this general structure from other migrations doing basically the same thing for other unique indexes ... but I think what's happening is that the interaction between using concurrent creation, strong_migrations and the disable ddl line and transactions, we wind up needing this approach to avoid that error, even though (like you are noting) it seems weird to redo something that the error would imply did not actually happen.

I can probably get a more detailed explanation if needed.

Reply:
What might happen here is that we are not in a transaction, so the index is created empty, then populated, and when the 2nd step fails because of duplicates, the empty index is left behind, hence needing to be deleted.

If we had other migrations doing it the same way then I will assume it works :)